[Hardware][CPU] Add ARM CPU backend

ShawnD200 opened the pull request:
Add ARM CPU backend support

(1) Based on the groundbreaking x86 CPU backend support, implemented the vector types on ARM neon
(2) Added tested cases for explicit implementation verification
```
# cd tests/cpu
# pip install .
# pytest . -v
```

Tested on MacOS with M2 Pro chip with docker image build, should work on generic ARM64/Linux.

Future work:
* Built from source on bare-metal MacOS
* Improve performance with advanced SIMD, and a lot other stuff
* Test on ARM/Linux and with more models

Tested simple models, just works. Have fun. Feedback is welcome.

ShawnD200 committed:
[Hardward][CPU] Add ARM CPU backend

Signed-off-by: Shawn Du <shawnd200@outlook.com>

ShawnD200 committed:
[Doc] Fix references

Signed-off-by: Shawn Du <shawnd200@outlook.com>

Isotr0py reviewed (commented):
Thanks for implementing arm64 support! The cpu kernel tests look pretty good! Just have some questions.

(I don't have an arm64 device. See if others having arm64 can run the test on native Linux)

ShawnD200 commented on requirements-arm.txt:
```diff
@@ -0,0 +1,8 @@
+# Common dependencies
+-r requirements-common.txt
+
+torch == 2.4.0
+torchvision  # required for the image processor of phi3v, this must be updated alongside torch
```
Honestly I haven't checked through the versions so that I don't know which is supported or not. This torch version is just copied from the x86 CPU backend. This could be future work.

ShawnD200 commented on csrc/cpu/cpu_types_arm.hpp:
```diff
@@ -0,0 +1,571 @@
+
+namespace vec_op {
+
+// FIXME: FP16 is not fully supported in Torch-CPU
+#define VLLM_DISPATCH_CASE_FLOATING_TYPES(...)                                 \
+  AT_DISPATCH_CASE(at::ScalarType::Float, __VA_ARGS__)                         \
+  AT_DISPATCH_CASE(at::ScalarType::BFloat16, __VA_ARGS__)
+
```
Yes, good point. X86 already added the switch case for Half type, ARM needs to add it too. This brings to a general question: how to automatically serve all ISA's, as a starter, those data types should be interfaces. I will add this to future work too. Thank you so much.

ShawnD200 committed:
Resolve merge conflicts

ShawnD200 committed:
1. Add fp16 support
2. Use compound vector types
3. Refactor cpu_types
4. Improve fma and storeFP32
5. Clean up

Signed-off-by: Shawn Du <shawnd200@outlook.com>

ShawnD200 committed:
Merge remote-tracking branch 'origin/main' into Add-Arm-CPU-backend

ShawnD200 commented:
Pushed a few updates:

Inspired by @Isotr0py, I realized there is tremendous complexity induced by combination of hardware supports and model inference features. Therefore, I think maybe it requires a generic layer that can abstract away the underlying hardware differences. To that end, here comes CPU types interfaces which can be implemented by various ISAs and features within ISA. I haven't sorted many things out; your feedback is so desired. 

BTW, before @mgoin introduced #9228 to me, which already implemented ARM CPU backend basically the same way, I had absolutely no idea of that PR. I started this work with a very simple goal that is to run vLLM on my macbook. Kudos to @sanketkaleoss, I learnt a lot from your work. Thanks.

ShawnD200 commented:
@mgoin @bigPYJ1151 @tlrmchlsmth @sanketkaleoss @Isotr0py, your feedback would be very helpful. Thanks.

ShawnD200 committed:
Break long lines and adjust imports

Signed-off-by: Shawn Du <shawnd200@outlook.com>

GaotangLi commented:
> Does this issue already being fix? I try install in vLLM and still got error for arm64

Did you resolve this?

syafiqmuda commented:
Sry I just change the Server to x64 instead of ARM.
